### PR TITLE
Save absolute memory limit alongside with percent

### LINKED ISSUE #111: Unexpected with -M
```
sudo earlyoom -M 1
// I expect 1 KiB threshold
earlyoom v1.2-19-ga48e825
mem total: 13855 MiB, swap total:    0 MiB
Sending SIGTERM when mem <=  0 % and swap <= 10 %,
        SIGKILL when mem <=  0 % and swap <=  5 %
mem avail: 12488 of 13855 MiB (90 %), swap free:    0 of    0 MiB ( 0 %)
mem avail: 12480 of 13855 MiB (90 %), swap free:    0 of    0 MiB ( 0 %)
mem avail: 12481 of 13855 MiB (90 %), swap free:    0 of    0 MiB ( 0 %)
mem avail: 12480 of 13855 MiB (90 %), swap free:    0 of    0 MiB ( 0 %)
mem avail: 12480 of 13855 MiB (90 %), swap free:    0 of    0 MiB ( 0 %)
mem avail: 12481 of 13855 MiB (90 %), swap free:    0 of    0 MiB ( 0 %)
mem avail: 12481 of 13855 MiB (90 %), swap free:    0 of    0 MiB ( 0 %)
mem avail: 12480 of 13855 MiB (90 %), swap free:    0 of    0 MiB ( 0 %)
...
mem avail:  390 of 13855 MiB ( 2 %), swap free:    0 of    0 MiB ( 0 %)
mem avail:  307 of 13855 MiB ( 2 %), swap free:    0 of    0 MiB ( 0 %)
mem avail:  224 of 13855 MiB ( 1 %), swap free:    0 of    0 MiB ( 0 %)
mem avail:  141 of 13855 MiB ( 1 %), swap free:    0 of    0 MiB ( 0 %)
Low memory! At or below SIGKILL limits (mem: 0 %, swap: 5 %)
mem avail:  **132 of 13855 MiB ( 0 %)**, swap free:    0 of    0 MiB ( 0 %)
Killing process 'python3' with signal 9, pid: 3543, badness: 890, VmRSS: 12313 MiB
mem avail: 12471 of 13855 MiB (90 %), swap free:    0 of    0 MiB ( 0 %)
mem avail: 12465 of 13855 MiB (89 %), swap free:    0 of    0 MiB ( 0 %)
mem avail: 12465 of 13855 MiB (89 %), swap free:    0 of    0 MiB ( 0 %)
```

- 132 MiB is 1%
- 132 MiB is not 1 KiB (it was set via -M 1)

I expect termination at 1 KiB, not killing at 132 MiB.

### PR DESCRIPTION
**Pull request by yangfl**
This also allow simultaneous use of -M and -m and change of memory/swap.

Fix #111

### CODE REVIEW

File: meminfo.h [hunk 1]
```diff
@@ -6,10 +6,12 @@
 
 #include <stdbool.h>
 
+typedef int percent_t;
```
Comment (rfjakob) [inline meminfo.h]: I think int is good enough

File: main.c [hunk 2]
```diff
@@ -95,17 +95,17 @@ int main(int argc, char* argv[])
             if (strlen(tuple.err)) {
                 fatal(15, "-m: %s", tuple.err);
             }
-            args.term.mem.percent = tuple.term;
-            args.kill.mem.percent = tuple.kill;
+            args.term.mem.percent = (percent_t)tuple.term;
+            args.kill.mem.percent = (percent_t)tuple.kill;
```
Comment (rfjakob) [inline main.c]: percent_t does not improve clarity I'm afraid

File: main.c [hunk 3]
```diff
@@ -95,17 +95,17 @@ int main(int argc, char* argv[])
...
         case 's':
             // Using "-s 100" is a valid way to ignore swap usage
             tuple = parse_term_kill_tuple(optarg, 100);
             if (strlen(tuple.err)) {
                 fatal(16, "-s: %s", tuple.err);
             }
-            args.term.swap.percent = tuple.term;
-            args.kill.swap.percent = tuple.kill;
+            args.term.swap.percent = (percent_t)tuple.term;
+            args.kill.swap.percent = (percent_t)tuple.kill;
```
Comment (rfjakob) [inline main.c]: percent_t does not improve clarity I'm afraid

File: main.c [hunk 4]
```diff
@@ -329,11 +329,11 @@ static int sleep_time_ms(const poll_loop_args_t* args, const meminfo_t* m)
     const int min_sleep = 100;
     const int max_sleep = 1000;
 
-    int mem_headroom_kib = (m->Mem.AvailablePercent - args->term.mem.percent) * 10 * m->Mem.Total / 1024;
+    int mem_headroom_kib = (m->Mem.AvailablePercent - args->term.mem.percent) * 10 * (int)(m->Mem.Total / 1024);
```
Comment (rfjakob) [inline main.c]: Use long long

File: main.c [hunk 5]
```diff
@@ -329,11 +329,11 @@ static int sleep_time_ms(const poll_loop_args_t* args, const meminfo_t* m)
...
     if (mem_headroom_kib < 0) {
         mem_headroom_kib = 0;
     }
-    int swap_headroom_kib = (m->Swap.AvailablePercent - args->term.swap.percent) * 10 * m->Swap.Total / 1024;
+    int swap_headroom_kib = (m->Swap.AvailablePercent - args->term.swap.percent) * 10 * (int)(m->Swap.Total / 1024);
```
Comment (rfjakob) [inline main.c]: Use long long

File: main.c [hunk 6]
```diff
@@ -388,7 +388,7 @@ static void poll_loop(const poll_loop_args_t args)
         }
         int sleep_ms = sleep_time_ms(&args, &m);
         debug("adaptive sleep time: %d ms\n", sleep_ms);
-        usleep(sleep_ms * 1000);
+        usleep((unsigned int)sleep_ms * 1000);
```
Comment (rfjakob) [inline main.c]: Use `(unsigned)`

File: meminfo.c [hunk 7]
```diff
@@ -92,9 +92,9 @@ meminfo_t parse_meminfo()
     }
 
     // Calculate percentages
-    m.Mem.AvailablePercent = m.Mem.Available * 100 / m.Mem.Total;
+    m.Mem.AvailablePercent = (percent_t)(m.Mem.Available * 100 / m.Mem.Total);
```
Comment (rfjakob) [inline meminfo.c]: Drop `percent_t`

File: meminfo.c [hunk 8]
```diff
@@ -92,9 +92,9 @@ meminfo_t parse_meminfo()
...
     if (m.Swap.Total > 0) {
-        m.Swap.AvailablePercent = m.Swap.Available * 100 / m.Swap.Total;
+        m.Swap.AvailablePercent = (percent_t)(m.Swap.Available * 100 / m.Swap.Total);
```
Comment (rfjakob) [inline meminfo.c]: Drop `percent_t`

File: meminfo.h [hunk 9]
```diff
@@ -6,10 +6,12 @@
...
+
 typedef struct {
     long long Total; // KiB
     long long Available; // -1 means no data available
-    int AvailablePercent; // percent of total memory that is available
+    percent_t AvailablePercent; // percent of total memory that is available
```
Comment (rfjakob) [inline meminfo.h]: int

File: kill.h [hunk 10]
```diff
@@ -5,13 +5,23 @@
 #include <regex.h>
 #include <stdbool.h>
 
+#include "meminfo.h"
+
+typedef struct {
+    int percent;
+    long long size;
+} capacity_limit_t;
```
Comment (rfjakob) [inline kill.h]: Naming suggestion: `limit_tuple_t` ?

File: main.c [hunk 11]
```diff
@@ -102,41 +95,33 @@ int main(int argc, char* argv[])
...
             }
-            args.swap_term_percent = tuple.term;
-            args.swap_kill_percent = tuple.kill;
-            have_s = 1;
+            args.term.swap.percent = tuple.term;
+            args.kill.swap.percent = tuple.kill;
             break;
         case 'M':
-            tuple = parse_term_kill_tuple(optarg, m.MemTotalKiB * 100 / 99);
+            tuple = parse_term_kill_tuple(optarg, 0);
```
Comment (rfjakob) [inline main.c]: 99% limit still makes sense, no?

**Comment by yangfl** [comment]
It does result in simpler assembly code (revealed by decompiler) and avoid two copy operations, and it is always good choice to pass by reference (address) since passing by value for structures is nearly always malformed in C/C++. But of course earlyoom is still fast enough even if it's 100x slower. So I didn't insist on this commit and you can have another opinion.

**Comment by rfjakob** [comment]
Hi, I implemented this in master another way.

About the pass-by-reference, we can do that, but then it should be a const pointer.

**Comment by rfjakob** [comment]
Looks like lgtm.com agrees with you :)
https://lgtm.com/projects/g/rfjakob/earlyoom/alerts/?mode=list

**Pull request closed without merging**